### DISCUSSION TITLE
Issues in running ansible playbook for adding virtual machine information

### DISCUSSION THREAD

**Discussion by AkhilaGarlapally** [netbox-community/ansible_modules#679] [Q&A] [answered]
HI,

I'm using ansible version:
```
 ansible --version
ansible 2.9.25
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.18 (default, Jun 10 2021, 00:11:02) [GCC 7.3.1 20180712 (Red Hat 7.3.1-13)]

```
pynetbox version is below```Package            Version
------------------ ---------
certifi            2021.10.8
charset-normalizer 2.0.9
idna               3.3
pip                20.1.1
pynetbox           6.4.1
requests           2.26.0
setuptools         47.1.0
six                1.16.0
urllib3            1.26.7
```
and playbook looks like below:
```- name: "Test NetBox modules"
  connection: local
  hosts: localhost
  gather_facts: False
  become: true
  become_user: true
  #plugin: netbox.netbox.netbox_virtual_machine
  vars:
    ansible_python_interpreter: /usr/bin/python3
  tasks:
    - name: return python interpreter
      debug:
        msg: "Python interpreter = {{ ansible_python_interpreter }}"
    - name: Create virtual machine within NetBox with only required information
      netbox.netbox.netbox_virtual_machine:
        netbox_url: https://netbox
        netbox_token: netbox_token
        data:
          name: Test
          cluster: test
        state: present
```
and the error im facing is below
```ansible-playbook 2.9.25
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible-playbook
  python version = 2.7.18 (default, Jun 10 2021, 00:11:02) [GCC 7.3.1 20180712 (Red Hat 7.3.1-13)]
Using /etc/ansible/ansible.cfg as config file
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Parsed /etc/ansible/hosts inventory source with ini plugin
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
Skipping callback 'actionable', as we already have a stdout callback.
Skipping callback 'counter_enabled', as we already have a stdout callback.
Skipping callback 'debug', as we already have a stdout callback.
Skipping callback 'dense', as we already have a stdout callback.
Skipping callback 'dense', as we already have a stdout callback.
Skipping callback 'full_skip', as we already have a stdout callback.
Skipping callback 'json', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'null', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.
Skipping callback 'selective', as we already have a stdout callback.
Skipping callback 'skippy', as we already have a stdout callback.
Skipping callback 'stderr', as we already have a stdout callback.
Skipping callback 'unixy', as we already have a stdout callback.
Skipping callback 'yaml', as we already have a stdout callback.

PLAYBOOK: main.yaml **********************************************************************************************************************************************************************************************
1 plays in main.yaml

PLAY [Test NetBox modules] ***************************************************************************************************************************************************************************************
META: ran handlers

TASK [return python interpreter] *********************************************************************************************************************************************************************************
task path: /root/main.yaml:11
ok: [localhost] => {
    "msg": "Python interpreter = /usr/bin/python3"
}

TASK [Create virtual machine within NetBox with only required information] ***************************************************************************************************************************************
task path: /root/main.yaml:14
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /var/tmp `"&& mkdir "` echo /var/tmp/ansible-tmp-1640277186.22-3484-213221559342289 `" && echo ansible-tmp-1640277186.22-3484-213221559342289="` echo /var/tmp/ansible-tmp-1640277186.22-3484-213221559342289 `" ) && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /var/tmp/ansible-tmp-1640277186.22-3484-213221559342289/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 147, in run
    res = self._execute()
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 665, in _execute
    result = self._handler.run(task_vars=variables)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/normal.py", line 47, in run
    result = merge_hash(result, self._execute_module(task_vars=task_vars, wrap_async=wrap_async))
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/__init__.py", line 825, in _execute_module
    (module_style, shebang, module_data, module_path) = self._configure_module(module_name=module_name, module_args=module_args, task_vars=task_vars)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/__init__.py", line 211, in _configure_module
    **become_kwargs)
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 1283, in modify_module
    environment=environment)
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 1120, in _find_module_utils
    py_module_cache, zf)
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 751, in recursive_finder
    [os.path.join(*py_module_name[:-idx])])
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 671, in __init__
    self.get_source()
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 686, in get_source
    pkg = import_module(self._package_name)
  File "/usr/lib64/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/usr/lib/python2.7/site-packages/ansible/utils/collection_loader.py", line 107, in load_module
    mod = self._find_module(fullname, None, load=True)[1]
  File "/usr/lib/python2.7/site-packages/ansible/utils/collection_loader.py", line 221, in _find_module
    code_object = compile(source=source, filename=source_path, mode='exec', flags=0, dont_inherit=True)
  File "/root/.ansible/collections/ansible_collections/netbox/netbox/plugins/module_utils/netbox_utils.py", line 886
    f"One or more of the kwargs provided are invalid for {parent}, provided kwargs: {', '.join(sorted(provided_kwargs))}. Acceptable kwargs: {', '.join(sorted(acceptable_query_params))}"
                                                                                                                                                                                         ^
SyntaxError: invalid syntax
fatal: [localhost]: FAILED! => {
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}

PLAY RECAP ***************************************************************************************************************````

Please help me resolve this issue

**Comment by sc68cal** [comment, marked as answer]
The Ansible that you have installed uses `Python 2.7.18` - see your output from `ansible --version` - setting `ansible_python_interpreter` does not fix this issue, since you can see in the stack trace that you are still loading code from the Python 2.7 installation.

In general, we do not support Python 2 when using this collection. I would suggest investigating something like [`pyenv`](https://github.com/pyenv/pyenv) or if your Linux distribution has Python 3 available, and an Ansible package that uses Python 3 as its dependency.

**Reply by AkhilaGarlapally** [reply]
Hi,

Thanks for your answer and yes i was able to resolve this issue with python3. But facing some other issues so far im using 
# /root/.ansible/collections/ansible_collections
Collection    Version
------------- -------
netbox.netbox 3.4.0

pynetbox           4.3.3

------------------ ---------
ansible            4.10.0

Below is the error im facing could you please help me resolve that issue 

> TASK [Create virtual machine within NetBox with only required information] ***************************************************************************************************************************************
> task path: /root/main.yaml:19
> <127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
> <127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
> <127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1640797969.4157438-3441-213307717754927 `" && echo ansible-tmp-1640797969.4157438-3441-213307717754927="` echo /root/.ansible/tmp/ansible-tmp-1640797969.4157438-3441-213307717754927 `" ) && sleep 0'
> Using module file /root/.ansible/collections/ansible_collections/netbox/netbox/plugins/modules/netbox_virtual_machine.py
> <127.0.0.1> PUT /root/.ansible/tmp/ansible-local-3406texo459i/tmp48nn23b6 TO /root/.ansible/tmp/ansible-tmp-1640797969.4157438-3441-213307717754927/AnsiballZ_netbox_virtual_machine.py
> <127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1640797969.4157438-3441-213307717754927/ /root/.ansible/tmp/ansible-tmp-1640797969.4157438-3441-213307717754927/AnsiballZ_netbox_virtual_machine.py && sleep 0'
> <127.0.0.1> EXEC /bin/sh -c '/root/my_app/env/bin/python3 /root/.ansible/tmp/ansible-tmp-1640797969.4157438-3441-213307717754927/AnsiballZ_netbox_virtual_machine.py && sleep 0'
> <127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1640797969.4157438-3441-213307717754927/ > /dev/null 2>&1 && sleep 0'
> The full traceback is:
>   File "/tmp/ansible_netbox.netbox.netbox_virtual_machine_payload_at9wdnlb/ansible_netbox.netbox.netbox_virtual_machine_payload.zip/ansible_collections/netbox/netbox/plugins/module_utils/netbox_utils.py", line 602, in _connect_netbox_api
>     self.version = nb.version
>   File "/root/my_app/env/lib64/python3.7/site-packages/pynetbox/api.py", line 230, in version
>     http_session=self.http_session,
>   File "/root/my_app/env/lib64/python3.7/site-packages/pynetbox/core/query.py", line 200, in get_version
>     raise RequestError(req)
> fatal: [localhost]: FAILED! => {
>     "changed": false,
>     "invocation": {
>         "module_args": {
>             "cert": null,
>             "data": {
>                 "cluster": "test",
>                 "comments": null,
>                 "custom_fields": null,
>                 "disk": null,
>                 "local_context_data": null,
>                 "memory": null,
>                 "name": "Test",
>                 "platform": null,
>                 "primary_ip4": null,
>                 "primary_ip6": null,
>                 "site": null,
>                 "status": null,
>                 "tags": null,
>                 "tenant": null,
>                 "vcpus": null,
>                 "virtual_machine_role": null
>             },
>             "netbox_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
>             "netbox_url": "https://netbox",
>             "query_params": null,
>             "state": "present",
>             "validate_certs": true
>         }
>     },
>     "msg": "Failed to establish connection to NetBox API"
> }
>

**Reply by sc68cal** [reply]
That's an error that occurs very early in the process of establishing a connection to NetBox, I would check that you have set all your connection details correctly, and also attempt to connect to NetBox via CURL or another HTTP utility. Sadly the code in pynetbox has a tendency to return very unhelpful errors.

**Reply by rodvand** [reply]
And if using self-signed certs, try setting `validate_certs: false`.

**Reply by AkhilaGarlapally** [reply]
Hi,

I tried with the curl request but im facing this error  curl -X GET -H "Authorization: Token xxx" https://netbox -H "accept: application/json"
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol is this something you could help me with??? does this playbook should be executed from the same server where netbox is running??

**Reply by sc68cal** [reply]
That indicates that you have something wrong with the server that NetBox is running on or the machine you are executing Ansible on.

The NetBox Ansible modules are executed on the Ansible controller (the machine that you invoke ansible-playbook on). So, you need to ensure that you are able to access the NetBox server from the machine you are invoking `ansible-playbook`

Common misconfiguration includes  things like firewall rules, nginx proxy configuration.

Are you able to reach the NetBox server vis your web browser?
